### PR TITLE
Normalize vectors

### PR DESCRIPTION
**Pull request by dimmu**
This version does the following things: 

1. Contains a model trained on the v0.2 revision of the taxonomy
2. Contains normalized vectorization for the hashing classifier
3. Simplifies the language specific workflows for the usermodel (since essentially we will be starting from scratch).

### CODE REVIEW

File: BUILD.gn [hunk 1]
```diff
@@ -49,20 +50,33 @@ source_set("usermodel") {
   output_name = "bat_native_usermodel"
 
   sources = [
+    "advertising_categories.h",
     "include/bat/usermodel/user_model.h",
-    "src/bag_of_words_extractor.h",
-    "src/bag_of_words_extractor.cc",
-    "src/naive_bayes.h",
-    "src/naive_bayes.cc",
+#    "src/bag_of_words_extractor.h",
```
Comment (tmancey) [inline BUILD.gn:55]: Can we please remove commented out sources

File: src/advertising_categories.h [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	advertising_categories["sports-volleyball"] = 186;
+	advertising_categories["sports-wrestling"] = 187;
+	advertising_categories["sports-yoga"] = 188;
+	advertising_categories["technology & computing-apple"] = 189;
+	advertising_categories["technology & computing-database"] = 190;
+	advertising_categories["technology & computing-freeware"] = 191;
+	advertising_categories["technology & computing-gaming"] = 192;
+	advertising_categories["technology & computing-programming"] = 193;
+	advertising_categories["technology & computing-software"] = 194;
+	advertising_categories["technology & computing-technology & computing"] = 195;
```
Comment (tmancey) [inline src/advertising_categories.h:200]: Are we intentionally missing macOS and others? Thanks

File: src/advertising_categories.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
+#include <map> 
+
+std::map<std::string, int> get_advertising_categories(){
+    std::map<std::string,int> advertising_categories = {};
```
Comment (tmancey) [inline src/advertising_categories.h:4]: nit:
```
std::map<std::string, int> advertising_categories = {
  {"architecture-architecture", 0},
  {"arts & entertainment-animation", 1},
  {"arts & entertainment-anime", 2},
  ...
};
```

File: src/classifier.cc [hunk 4]
```diff
@@ -0,0 +1,50 @@
+#include "classifier.h"
```
Comment (tmancey) [inline src/classifier.cc:6]: Missing copyright notice

File: src/classifier.cc [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    //third pass: divide by sum_exp
+    for (auto const& x : rtn){
+        rtn[x.first] = rtn[x.first]/sum_exp;       
+    }
+
+    return rtn;
+}
+Linear_classifier::Linear_classifier(){}
+
+Linear_classifier::Linear_classifier(std::map<std::string, DataPoint> weights_, std::map<std::string,float> biases_){
```
Comment (tmancey) [inline src/classifier.cc]: underscore should only be used for member variables

File: src/classifier.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#include <iostream>
```
Comment (tmancey) [inline src/classifier.h:6]: Missing copyright notice

Comment (tmancey) [reply src/classifier.h:6]: Missing header guards

File: src/classifier.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+#include "data_point.h"
+#include <vector>
+namespace usermodel{
+
+// class Classifier{
```
Comment (tmancey) [inline src/classifier.h]: nit: remove commented out code

File: BUILD.gn [hunk 8]
```diff
@@ -1,6 +1,7 @@
-# Copyright 2014 The Chromium Authors. All rights reserved.
-# Use of this source code is governed by a BSD-style license that can be
-# found in the LICENSE file.
+# Copyright (c) 2018 The Brave Authors. All rights reserved.
```
Comment (tmancey) [inline BUILD.gn]: 2020 not 2018

File: README.md [hunk 9]
```diff
@@ -80,12 +80,13 @@ supercategory of "sports" with a subcategory of "rugby".
...
 
-Initialize the User Model by calling `InitializePageClassifier(model)` where
+Initialize the User Model by calling `InitializePageClassifier(model, region)`
+where `region` is a `std::string` containing the region of the user model and
 `model` is a `std::string` containing the contents of the `model.json` file as
 follows:
 
 ```
-bool success = InitializePageClassifier(model)
+bool success = InitializePageClassifier(model, region)
```
Comment (tmancey) [inline README.md:89]: `region` after our discussion should be `language_code`

File: TODO.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
+# Plan:
```
Comment (tmancey) [inline TODO.md:1]: Discussed and should probably be updated

File: src/classifier.cc [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    weights = weights_;
+    biases = biases_;
+}
+Linear_classifier::Linear_classifier(Linear_classifier &other){
+    weights = other.weights;
+    biases = other.biases;
+}
+Linear_classifier::~Linear_classifier()=default;
+std::map<std::string, float> Linear_classifier::Predict(DataPoint x){
```
Comment (tmancey) [inline src/classifier.cc:46]: const reference

File: src/classifier.h [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+//         ~Classifier();
+//         std::map<std::string, float> Predict(Data_point x);
+// };
+std::map<std::string, float> Softmax( std::map<std::string, float> y);
+
+class Linear_classifier{
+    public:
+        Linear_classifier();
+        Linear_classifier(Linear_classifier &other);
+        Linear_classifier(std::string model);
```
Comment (tmancey) [inline src/classifier.h]: const std::string& model

File: src/data_point.h [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        DataType get_type();
+        int get_dims();
+        DataType type;
+        int n_dims;
+        std::string data_text;
+        std::vector<float> data_vector;
+        std::map<unsigned,float> data_sparse;
+};
+
+// float inner_product(DataPoint a, DataPoint b);
```
Comment (tmancey) [inline src/data_point.h]: nit: commented out code

File: src/data_point.h [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+#ifndef DATAPOINT_H
+#define DATAPOINT_H
+#include <map>
+#include <vector>
+#include <iostream>
+namespace usermodel{
+
+enum class DataType{
+    text_data = 0,
```
Comment (tmancey) [inline src/data_point.h]: nit: `kTextData` etc.

File: src/data_point.cc [hunk 15]
```diff
@@ -0,0 +1,66 @@
+#include "data_point.h"
+#include <stdexcept>
```
Comment (tmancey) [inline src/data_point.cc:7]: std lib #include's should be first

File: src/hashing_extractor.cc [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+#include <algorithm>
+#include "CRC.h"
+#include "static_values.h"
+#include <vector>
+
+namespace usermodel {
+
+HashVectorizer::~HashVectorizer() = default;
+HashVectorizer::HashVectorizer(){
+  for (int i = 1; i<=6;i++)
```
Comment (tmancey) [inline src/hashing_extractor.cc]: nit: can we define 6 in static_values.h?

File: src/hashing_extractor.cc [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    substring_sizes.push_back(i); 
+  num_buckets = 10000;
```
Comment (tmancey) [inline src/hashing_extractor.cc]: nit: can we define 10000 in static_values.h?

File: src/hashing_extractor.cc [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+}
+HashVectorizer::HashVectorizer(int n_buckets, std::vector<int> subgrams){
+  for (unsigned i = 0; i<subgrams.size();i++)
```
Comment (tmancey) [inline src/hashing_extractor.cc]: nit: should use size_t instead of unsigned

File: src/hashing_extractor.cc [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    std::wstring_convert<std::codecvt_utf8<wchar_t>> conv1;
+    std::string u8str = conv1.to_bytes(substring);
+    auto rtn = CRC::Calculate(u8str.data(), u8str.size(), CRC::CRC_32()) % num_buckets; 
+  return rtn;
+}
+
+std::map<unsigned, float> HashVectorizer::GetFrequencies(const std::wstring& html) {
+  std::wstring data = html;
+  std::map<unsigned, float> frequencies;
+  frequencies = {};
```
Comment (tmancey) [inline src/hashing_extractor.cc]: nit: not required as container is initialized above on line 44

File: src/hashing_extractor.cc [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+  // TODO: Review maximum length
```
Comment (tmancey) [inline src/hashing_extractor.cc]: Does this still need doing?

File: src/hashing_extractor.h [hunk 21]
```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (tmancey) [inline src/hashing_extractor.h]: Wrong copyright

File: src/hashing_extractor.h [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+#include <map>
+#include <vector>
+
+namespace usermodel {
+
+class HashVectorizer {
+ public:
+  HashVectorizer();
+  HashVectorizer(const HashVectorizer& other);
+  HashVectorizer(int n_buckets, std::vector<int> subgrams);
```
Comment (tmancey) [inline src/hashing_extractor.h:19]: const otherwise a copy occurs

File: src/pipeline.cc [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+namespace usermodel {
+
+Pipeline::Pipeline(){
+    version_ = 0;
+    timestamp_ ="";
+    locale_ = "en";
+    transformations_ = {};
+    advertising_categories_ = get_advertising_categories();
+    get_reverse_categories();
+    // classifier_ = nullptr;
```
Comment (tmancey) [inline src/pipeline.cc:20]: ?

File: src/pipeline.cc [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+  for (size_t i = 0 ; i < specified_classes->GetList().size();i++){
+    const base::Value& class_name = specified_classes->GetList()[i];
+    classes.push_back(class_name.GetString());
+  }
+  base::Value* class_weights = classifier->FindKey("class_weights");
+  if ( !class_weights->is_dict()) {
+    return false;
+  }
+  std::map<std::string,DataPoint> weights;
+  weights = {};
```
Comment (tmancey) [inline src/pipeline.cc]: not needed, please check elsewhere

File: src/pipeline.cc [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+    std::vector<float> tmp_weights = {};
+    for (size_t j = 0 ; j < this_class->GetList().size(); j++ ){
+      const base::Value& weight = this_class->GetList()[j];
+      tmp_weights.push_back( static_cast<float> (weight.GetDouble()) );
+    }
+    auto dubs = usermodel::DataPoint(tmp_weights);
+    weights.insert({classes[i], dubs});
+  }
+  //finally do the biases: 
+  std::map<std::string, float> specified_biases = {};
```
Comment (tmancey) [inline src/pipeline.cc:195]: nit: `= {};` not required

File: src/pipeline.cc [hunk 26]
```diff
@@ -0,0 +1,266 @@
...
+  base::Value* biases = classifier->FindKey("biases");
+  if (!biases->is_list()){
+    return false;
+  }
+  if (biases->GetList().size() != classes.size()){
+    return false;
+  }
+  for (size_t i = 0 ; i < biases->GetList().size();i++){
+    const base::Value& this_bias = biases->GetList()[i];
+    specified_biases.insert({classes[i], static_cast<float> (this_bias.GetDouble())}) ;
```
Comment (tmancey) [inline src/pipeline.cc]: nit: use classes.at(i) instead of classes[i]

File: src/transformation.cc [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+    auto norm = sqrt(s);
+    //divide original datapoint entries by the norm
+    for (auto const& x : datapoint.data_sparse){
+        normalized_vector[x.first]= x.second / norm;
+    }
+    return DataPoint(normalized_vector,hash_vectorizer.get_buckets());
+}
+///////////////////////////////////////////////////////////
+
+To_lower::To_lower(){type = TransformationType::TO_LOWER;}
```
Comment (tmancey) [inline src/transformation.cc:66]: use `base::ToLowerASCII`

File: src/transformation.cc [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+
+DataPoint Transformation::get(DataPoint inp){
+    switch(type){
+        case TransformationType::TO_LOWER:
+            return get_lower(inp);
+        case TransformationType::HASHED_NGRAMS:
+            return get_ngrams(inp);
+        case TransformationType::NORMALIZE:
+            return get_normalized(inp);
+        default:
```
Comment (tmancey) [inline src/transformation.cc]: For type safety we should remove default, add NONE if needed

File: src/transformation.h [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+    public:
+        Transformation();
+        Transformation(TransformationType t);
+        Transformation(std::string parameters);
+        ~Transformation();
+        DataPoint get(DataPoint inp);
+        TransformationType getType();
+    protected:
+        HashVectorizer hash_vectorizer;
+        TransformationType type;
```
Comment (tmancey) [inline src/transformation.h:31]: Not initialized in release builds, so will cause issues. Discussed adding a new type (first entry) and setting to NONE. Please check all other simple types, as they need initializing

File: src/advertising_categories.h (hunk 2 above)
Comment (dimmu) [reply src/advertising_categories.h:200]: Those were the exact 203 categories specified in the previous model under: 
`bat-native-ads/resources/user_models/languages/en/user_model.json`. This one was made to ensure there was a temporary parity between the two.

File: src/advertising_categories.h (hunk 3 above)
Comment (dimmu) [reply src/advertising_categories.h:4]: I would like to nuke this file entirely and just return key-values when this is merged, also referring to https://github.com/brave-intl/bat-native-usermodel/pull/25#discussion_r390191221

**Comment by dimmu** [comment]
Fixing the following: 
- Single predictions returned by classifier
- Byte level hashing extraction

**Pull request merged**